### ISSUE TITLE
Allow using proxy for ssh connection (sshconfig ProxyCommand)

### ISSUE THREAD

**Issue by m42e** [boringproxy/boringproxy#163]
It would be nice to have a possibility to use a proxy for the ssh connection (ProxyCommand corkscrew .....,  for example)

**Comment by anderspitman** [comment]
Hey @m42e, can you give an example of what you're trying to accomplish?

If you want to be able to SSH into machines where the boringproxy client is running, you can follow the instructions here:

https://forum.indiebits.io/t/how-to-forward-ssh-port-example-just-in-time-feature-request/56/5?u=anders

I believe you should be able to use it as a jump host as well.

Note that this approach is a bit clunky. I want to improve boringproxy's support for SSH in the future.

**Comment by m42e** [comment]
Sure, @anderspitman.
Sorry of the very brief description. 

So sometimes I am in an environment that does only allow connections to the outside world using an http proxy.
I am using corkscrew to do SSH over the http proxy. This is achieved by adding the line

```
ProxyCommand /usr/bin/corkscrew proxy_ip proxy_port %h %p
```

into the ssh_config file.

So it is still for a connection from the client to the server. To create the tunnel.

**Comment by anderspitman** [comment]
This is a very interesting use case. Basically for you to be able to use corkscrew, boringproxy would need to implement a standard HTTP proxy (ie using the HTTP CONNECT method). I'm unlikely to implement this.

However, with the way boringproxy currently works, I think it might be easy to support your use case. I believe all you need is a tiny program that listens on a local port, then tunnels any connections on that port through TLS to the boringproxy server. If your tunnel is correctly configured to point at the SSH server on the boringproxy client machine, it should work.

I feel like stunnel should support this, but everything I see indicates stunnel requires it's own special server. If not stunnel it seems like such a program must already exist.

In any case it should be easy to write. I think you just derailed what I was planning to spend the next couple hours doing ;)

**Comment by anderspitman** [comment]
It turned out to be just about as simple as expected! Here's the program:

**Comment by anderspitman** [comment]
Ok I believe I've fixed the TLS cert bug, but you'll need to use a master build:

https://gemdrive.io/apps/delver/?drive=https://files.apitman.com&path=/public/boringproxy/master

**Comment by m42e** [comment]
Thanks a lot @anderspitman. Especially I did not want to keep you from  anything important :)

Now I did not get correctly what you explained or I did not manage to explain my scenario exactly. I try again to explain:

![image](https://user-images.githubusercontent.com/2410802/155464749-d9f06b9e-e9f1-4950-a31c-8b4e2f366879.png)

On the PC1, boringproxy client is able to connect to the server using http. But it could not directly connect to some outside ssh port (nor can it with http, but there the proxy is respected).

For my normal ssh commands I added corkscrew to enable the green path (e.g. GitHub ssh clone)

And I really have the feeling I owe you at least a beer for your effort and immediate action. So you are welcome if you are in Munich at some point in time.

**Comment by anderspitman** [comment]
Nice diagram! I'm still not sure what the goal is. Are you trying to SSH into PC2 while you're at work, or are you trying to SSH into PC1 while you're at home? Or something else?

No worries about the effort. The raw TLS tunnel is a great feature. In fact I think I'm going to add a new command to boringproxy to support it. Basically it lets you tunnel any TCP program through boringproxy, very similar to what stunnel does. But that doesn't mean we can't hang out if I make my way to Muenchen one of these days!

**Comment by m42e** [comment]
Ok, I want to use boringproxy for sharing some html files (using a simple stupid python3 Webserver).

so PC2 shall Access the websites on PC1 using the boringproxy :)

**Comment by anderspitman** [comment]
Ah yes in that case the client's outgoing ssh connection on port 22 will be blocked. That said, if you can run the boringproxy server OpenSSH server on a different port that isn't blocked (maybe use port 80 for SSH and 443 for HTTPS) you might be able to get through. Someday I'd like to support a protocol that works over HTTP or WebSockets.

**Comment by anderspitman** [comment]
Also, I've [added](https://github.com/boringproxy/boringproxy/pull/166) the `boringproxy tuntls` command so we can now do raw TLS tunneling. Thanks for the idea!

**Comment by m42e** [comment]
Ok, I thought there might be a way to use the proxy for ssh, too. Maybe I can dive into go at some point in time.

**Comment by anderspitman** [comment]
Yeah unfortunately since boringproxy leverages OpenSSH for the heavy lifting, it needs to run on a separate port. Though unless I'm mistaken, corkscrew won't solve your problem either, will it?

Maybe take a look at these:

https://github.com/mmatczuk/go-http-tunnel (HTTP/2)

https://github.com/erebe/wstunnel (WebSockets)

**Comment by m42e** [comment]
It will and does, the line in corkscrew posted above, tunnels the outgoing ssh connection through the proxy.

**Comment by anderspitman** [comment]
Ohh I see. You're using corkscrew together with boringproxy.

**Comment by m42e** [comment]
That was the Idea, to let boringproxys ssh part use corkscrew. At the moment my native ssh client is using it.

Or does boring proxy utilise the ssh executable?

**Comment by anderspitman** [comment]
No the client uses the Golang ssh client. Unfortunately I don't think this will work without changes to the boringproxy client to tell it how to map the tunnel SSH port to the corkscrew port.

**Comment by m42e** [comment]
I do not have any experience but I stumbled across https://github.com/appleboy/easyssh-proxy
But I am not sure this is an option for boring proxy. It is fine for me if you say no.

**Comment by anderspitman** [comment]
Not sure. How are you thinking we might use it?

So, I opened a new issued (https://github.com/boringproxy/boringproxy/issues/167). I think it might take minimal changes to support tunneling the client's SSH connection through TLS, for situations like yours. It's pretty hacky but I'll think on it.

**Comment by anderspitman** [comment]
@m42e I may have gotten a bit carried away. I think boringproxy master now has everything you need to make this work. It takes a little manual work to set it up. Here's what you need to do:

Set up a tunnel named something like ssh.example.com with the following settings:
* Tunnel Port: 22
* Client Name: No Client
* TLS Termination: Server raw TLS

Then on the client machine, run the following:

`boringproxy tuntls -server ssh.example.com -port 2022`

Then whenever you create a tunnel do it as you normally would, but set the following:

* Override SSH Server Address: localhost
* Override SSH Server Port: 2022

And that should be it. Basically what this is doing is creating a proxy similar to corkscrew on the local port 2022, and telling the client to use localhost:2022 when connecting to the SSH server for the tunnels.

Let me know if you get a chance to try it out!

https://gemdrive.io/apps/delver/?drive=https://files.apitman.com&path=/public/boringproxy/master

**Comment by anderspitman** [comment]
And don't feel bad about the extra work. These features weren't too invasive, and I got to do some nice cleanup in the process.

**Comment by fopina** [comment]
I'm planning on a PR to add proxy support on SSH.

HTTP calls to the REST API already respect http_proxy/https_proxy so I'm adding support for that on the SSH connections as well using `golang.org/x/net/proxy`

Only thing that took me a bit longer to figure out was to be able to use HTTP CONNECT with that setup as it only supports SOCKS5 out of the box, but I have a PoC working now, so I'll be submitting a PR soon, I hope it still helps

**Comment by fopina** [comment]
as I need to use SSH over an HTTP proxy at work, I've created https://github.com/boringproxy/boringproxy/pull/183 to tackle it. I believe it also solves this issue

**Comment by wzwg3** [comment]
Hey, tried to follow your advice to open a TLS tunnel on port 2022 and route all client SSH tunnels through localhost:2022.

Opening a tunnel with
```
boringproxy tuntls -server ssh.proxy.<domain> -port 2022
```
I get `Listening on port 2022` - seems OK.

Then running the boringproxy client with a tunnel that uses `"server_address":"localhost","server_port":2022` we get
`BoreTunnel error:  Failed to dial: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain`.

A similar tunnel without server_address, server_port works perfectly.
We use a fresh build from the current sources.

Any hints whats going wrong here?
Thanks.

**Comment by anderspitman** [comment]
Hey @wzwg3. The tuntls functionality is not heavily tested. I just hacked it together since it seemed like a tool that might be useful. You say a normal tunnel is working ok? Is there a reason you want to use tuntls in that case? Can you clarify your overall goal?

Thanks!

**Comment by wzwg3** [comment]
We have a firewall that blocks SSH traffic. So the idea was to use a TLS tunnel between client and server (the tuntls tunnel) and route the SSH tunnel between boringproxy client and server through this tunnel.

**Comment by m42e** [comment]
I have created a pr that allows what I initially wanted to solve #212
